File: py-lynxes/src/lynxes/_csv.py
"""CSV reader front-end for Lynxes Python APIs."""

from lynxes._lynxes import NodeFrame
from lynxes._lynxes import read_csv_native_py as _read_csv_native


def _require_pyarrow():
    try:
        import pyarrow as pa
        import pyarrow.compute as pc
        import pyarrow.csv as csv
    except ImportError as exc:
        raise ImportError(
            "lynxes.read_csv(engine='pyarrow') requires pyarrow. Install pyarrow to use it."
        ) from exc

    return pa, pc, csv


def _string_array_from_column(column, *, name):
    pa, pc, _ = _require_pyarrow()
    if column.null_count:
        raise ValueError(f"{name} cannot contain null values")
    return pc.cast(column.combine_chunks(), pa.string())


def _label_array(label, rows):
    pa, _, _ = _require_pyarrow()
    offsets = pa.array(range(rows + 1), type=pa.int32())
    values = pa.array([label] * rows, type=pa.string())
    return pa.ListArray.from_arrays(offsets, values)


def _singleton_label_array(column, *, name):
    pa, pc, _ = _require_pyarrow()
    if column.null_count:
        raise ValueError(f"{name} cannot contain null values")
    values = pc.cast(column.combine_chunks(), pa.string())
    offsets = pa.array(range(len(values) + 1), type=pa.int32())
    return pa.ListArray.from_arrays(offsets, values)


def _csv_table_to_node_frame(table, *, label=None, id_col=None, id_prefix=None, columns=None):
    pa, _, _ = _require_pyarrow()
    column_names = table.column_names
    rows = table.num_rows

    if id_col is not None:
        if id_col not in column_names:
            raise ValueError(f"id_col {id_col!r} not found in CSV columns")
        id_array = _string_array_from_column(table[id_col], name=id_col)
    elif "_id" in column_names:
        id_array = _string_array_from_column(table["_id"], name="_id")
    else:
        prefix = id_prefix or "row"
        id_array = pa.array((f"{prefix}_{idx}" for idx in range(rows)), type=pa.string())

    if label is not None:
        label_array = _label_array(label, rows)
    elif "_label" in column_names:
        label_array = _singleton_label_array(table["_label"], name="_label")
    else:
        raise ValueError("read_csv requires label=... unless the CSV contains a _label column")

    arrays = [id_array, label_array]
    names = ["_id", "_label"]
    output_names = columns if columns is not None else column_names
    for name in output_names:
        if name in {"_id", "_label"}:
            continue
        if name.startswith("_"):
            raise ValueError(f"CSV column {name!r} is reserved by NodeFrame")
        if name not in column_names:
            raise ValueError(f"column {name!r} not found in CSV columns")
        arrays.append(table[name].combine_chunks())
        names.append(name)

    batch = pa.RecordBatch.from_arrays(arrays, names=names)
    return NodeFrame.from_arrow(batch)


def read_csv(
    path,
    *,
    label=None,
    id_col=None,
    id_prefix=None,
    columns=None,
    schema_overrides=None,
    engine="native",
    infer_schema_rows=None,
    batch_size=65536,
    has_header=True,
    delimiter=",",
    read_options=None,
    parse_options=None,
    convert_options=None,
):
    """Read a CSV file into a NodeFrame."""
    if engine == "native":
        if read_options is not None or parse_options is not None or convert_options is not None:
            raise ValueError("pyarrow CSV options require engine='pyarrow'")
        return _read_csv_native(
            path,
            label=label,
            id_col=id_col,
            id_prefix=id_prefix,
            columns=columns,
            schema_overrides=schema_overrides,
            infer_schema_rows=infer_schema_rows,
            batch_size=batch_size,
            has_header=has_header,
            delimiter=delimiter,
        )

    if engine != "pyarrow":
        raise ValueError("engine must be 'native' or 'pyarrow'")
    if infer_schema_rows is not None or batch_size != 65536 or not has_header or delimiter != ",":
        raise ValueError("native CSV options require engine='native'")

    pa, _, csv = _require_pyarrow()
    if convert_options is None:
        include_columns = _pyarrow_include_columns(columns, id_col=id_col, label=label)
        column_types = _pyarrow_schema_overrides(pa, schema_overrides)
        if include_columns is not None or column_types:
            convert_options = csv.ConvertOptions(
                include_columns=include_columns,
                column_types=column_types,
            )
    elif columns is not None or schema_overrides:
        raise ValueError(
            "columns/schema_overrides cannot be combined with explicit pyarrow convert_options"
        )
    table = csv.read_csv(
        path,
        read_options=read_options,
        parse_options=parse_options,
        convert_options=convert_options,
    )
    return _csv_table_to_node_frame(
        table,
        label=label,
        id_col=id_col,
        id_prefix=id_prefix,
        columns=columns,
    )


def _pyarrow_include_columns(columns, *, id_col, label):
    if columns is None:
        return None

    include_columns = []
    if id_col is not None:
        include_columns.append(id_col)
    elif "_id" in columns:
        include_columns.append("_id")
    if label is None:
        include_columns.append("_label")
    for name in columns:
        if name not in include_columns:
            include_columns.append(name)
    return include_columns


def _pyarrow_schema_overrides(pa, schema_overrides):
    if not schema_overrides:
        return {}

    marker_to_type = {
        "String": pa.string(),
        "StringView": getattr(pa, "string_view", pa.string)(),
        "Utf8View": getattr(pa, "string_view", pa.string)(),
        "Int": pa.int64(),
        "Float": pa.float64(),
        "Bool": pa.bool_(),
    }
    out = {}
    for name, dtype in schema_overrides.items():
        try:
            out[name] = marker_to_type[dtype]
        except KeyError as exc:
            raise ValueError(f"unsupported schema override dtype for {name!r}: {dtype!r}") from exc
    return out


def _node_frame_read_csv(
    _cls,
    path,
    *,
    label=None,
    id_col=None,
    id_prefix=None,
    columns=None,
    schema_overrides=None,
    engine="native",
    infer_schema_rows=None,
    batch_size=65536,
    has_header=True,
    delimiter=",",
    read_options=None,
    parse_options=None,
    convert_options=None,
):
    return read_csv(
        path,
        label=label,
        id_col=id_col,
        id_prefix=id_prefix,
        columns=columns,
        schema_overrides=schema_overrides,
        engine=engine,
        infer_schema_rows=infer_schema_rows,
        batch_size=batch_size,
        has_header=has_header,
        delimiter=delimiter,
        read_options=read_options,
        parse_options=parse_options,
        convert_options=convert_options,
    )


NodeFrame.read_csv = classmethod(_node_frame_read_csv)

__all__ = ["read_csv"]

File: crates/lynxes-io/src/io/csv.rs
use std::{fs::File, path::Path, sync::Arc};

use arrow::{
    array::{
        new_empty_array, Array, ArrayRef, ListBuilder, RecordBatch, StringArray, StringBuilder,
    },
    csv::reader::{Format, ReaderBuilder},
    datatypes::{DataType, Field, Schema as ArrowSchema},
};
use lynxes_core::{GFError, NodeFrame, Result, COL_NODE_ID, COL_NODE_LABEL};

const DEFAULT_BATCH_SIZE: usize = 65_536;

#[derive(Debug, Clone)]
pub struct CsvNodeReadOptions {
    pub label: Option<String>,
    pub id_col: Option<String>,
    pub id_prefix: Option<String>,
    pub columns: Option<Vec<String>>,
    pub schema_overrides: Vec<(String, DataType)>,
    pub infer_schema_rows: Option<usize>,
    pub batch_size: usize,
    pub has_header: bool,
    pub delimiter: u8,
}

impl Default for CsvNodeReadOptions {
    fn default() -> Self {
        Self {
            label: None,
            id_col: None,
            id_prefix: None,
            columns: None,
            schema_overrides: Vec::new(),
            infer_schema_rows: None,
            batch_size: DEFAULT_BATCH_SIZE,
            has_header: true,
            delimiter: b',',
        }
    }
}

pub fn read_csv_nodes(path: impl AsRef<Path>, options: &CsvNodeReadOptions) -> Result<NodeFrame> {
    let path = path.as_ref();
    let format = Format::default()
        .with_header(options.has_header)
        .with_delimiter(options.delimiter);

    let mut infer_file = File::open(path)?;
    let (source_schema, _) = format
        .infer_schema(&mut infer_file, options.infer_schema_rows)
        .map_err(|err| GFError::InvalidConfig {
            message: format!("failed to infer CSV schema for {}: {err}", path.display()),
        })?;
    let source_schema = apply_schema_overrides(source_schema, &options.schema_overrides)?;
    let projection = projection_indices(&source_schema, options)?;
    let read_schema = projected_schema(&source_schema, projection.as_deref());

    let read_file = File::open(path)?;
    let mut reader_builder = ReaderBuilder::new(Arc::new(source_schema.clone()))
        .with_format(format)
        .with_batch_size(options.batch_size.max(1));
    if let Some(projection) = projection {
        reader_builder = reader_builder.with_projection(projection);
    }
    let reader = reader_builder
        .build(read_file)
        .map_err(|err| GFError::InvalidConfig {
            message: format!("failed to open CSV reader for {}: {err}", path.display()),
        })?;

    let batches = reader
        .collect::<std::result::Result<Vec<_>, _>>()
        .map_err(|err| GFError::InvalidConfig {
            message: format!("failed to read CSV rows from {}: {err}", path.display()),
        })?;
    let source_columns = concat_source_columns(&read_schema, &batches)?;
    let rows = source_columns
        .first()
        .map(|array| array.len())
        .unwrap_or_else(|| batches.iter().map(RecordBatch::num_rows).sum());

    let source_names: Vec<&str> = read_schema
        .fields()
        .iter()
        .map(|field| field.name().as_str())
        .collect();

    let id_array = build_id_array(&read_schema, &source_columns, rows, options)?;
    let label_array = build_label_array(&read_schema, &source_columns, rows, options)?;

    let mut fields = vec![
        Field::new(COL_NODE_ID, DataType::Utf8, false),
        Field::new(
            COL_NODE_LABEL,
            DataType::List(Arc::new(Field::new("item", DataType::Utf8, true))),
            false,
        ),
    ];
    let mut columns = vec![id_array, label_array];

    append_user_columns(
        &mut fields,
        &mut columns,
        &read_schema,
        &source_columns,
        &source_names,
        options.columns.as_deref(),
    )?;

    let batch = RecordBatch::try_new(Arc::new(ArrowSchema::new(fields)), columns)
        .map_err(std::io::Error::other)?;
    NodeFrame::from_record_batch(batch)
}

fn apply_schema_overrides(
    schema: ArrowSchema,
    overrides: &[(String, DataType)],
) -> Result<ArrowSchema> {
    if overrides.is_empty() {
        return Ok(schema);
    }

    let mut fields: Vec<Field> = schema
        .fields()
        .iter()
        .map(|field| field.as_ref().clone())
        .collect();
    for (name, dtype) in overrides {
        let idx = schema.index_of(name).map_err(|_| GFError::ColumnNotFound {
            column: name.clone(),
        })?;
        fields[idx] = Field::new(name, dtype.clone(), true);
    }

    Ok(ArrowSchema::new(fields))
}

fn projection_indices(
    schema: &ArrowSchema,
    options: &CsvNodeReadOptions,
) -> Result<Option<Vec<usize>>> {
    let Some(requested_columns) = options.columns.as_deref() else {
        return Ok(None);
    };

    let mut names = Vec::new();
    if let Some(id_col) = options.id_col.as_deref() {
        names.push(id_col);
    } else if schema.index_of(COL_NODE_ID).is_ok() {
        names.push(COL_NODE_ID);
    }
    if options.label.is_none() {
        names.push(COL_NODE_LABEL);
    }
    for name in requested_columns {
        names.push(name);
    }

    let mut projection = Vec::new();
    for name in names {
        let idx = schema.index_of(name).map_err(|_| GFError::ColumnNotFound {
            column: name.to_owned(),
        })?;
        if !projection.contains(&idx) {
            projection.push(idx);
        }
    }

    Ok(Some(projection))
}

fn projected_schema(schema: &ArrowSchema, projection: Option<&[usize]>) -> ArrowSchema {
    let Some(projection) = projection else {
        return schema.clone();
    };

    let fields: Vec<Field> = projection
        .iter()
        .map(|idx| schema.field(*idx).clone())
        .collect();
    ArrowSchema::new(fields)
}

fn append_user_columns(
    fields: &mut Vec<Field>,
    columns: &mut Vec<ArrayRef>,
    schema: &ArrowSchema,
    source_columns: &[ArrayRef],
    source_names: &[&str],
    requested_columns: Option<&[String]>,
) -> Result<()> {
    if let Some(requested_columns) = requested_columns {
        let mut emitted = Vec::new();
        for name in requested_columns {
            let name = name.as_str();
            if name == COL_NODE_ID || name == COL_NODE_LABEL || emitted.contains(&name) {
                continue;
            }
            if name.starts_with('_') {
                return Err(GFError::ReservedColumnName {
                    column: name.to_owned(),
                });
            }
            let idx = schema.index_of(name).map_err(|_| GFError::ColumnNotFound {
                column: name.to_owned(),
            })?;
            fields.push(schema.field(idx).clone());
            columns.push(source_columns[idx].clone());
            emitted.push(name);
        }
        return Ok(());
    }

    for (idx, name) in source_names.iter().enumerate() {
        if *name == COL_NODE_ID || *name == COL_NODE_LABEL {
            continue;
        }
        if name.starts_with('_') {
            return Err(GFError::ReservedColumnName {
                column: (*name).to_owned(),
            });
        }
        fields.push(schema.field(idx).clone());
        columns.push(source_columns[idx].clone());
    }

    Ok(())
}

fn concat_source_columns(schema: &ArrowSchema, batches: &[RecordBatch]) -> Result<Vec<ArrayRef>> {
    let mut columns = Vec::with_capacity(schema.fields().len());

    for col_idx in 0..schema.fields().len() {
        if batches.is_empty() {
            columns.push(new_empty_array(schema.field(col_idx).data_type()));
            continue;
        }

        let arrays: Vec<ArrayRef> = batches
            .iter()
            .map(|batch| batch.column(col_idx).clone())
            .collect();
        let refs: Vec<&dyn Array> = arrays.iter().map(|array| array.as_ref()).collect();
        let array = arrow::compute::concat(&refs).map_err(std::io::Error::other)?;
        columns.push(array);
    }

    Ok(columns)
}

fn build_id_array(
    schema: &ArrowSchema,
    source_columns: &[ArrayRef],
    rows: usize,
    options: &CsvNodeReadOptions,
) -> Result<ArrayRef> {
    if let Some(id_col) = options.id_col.as_deref() {
        let idx = schema
            .index_of(id_col)
            .map_err(|_| GFError::ColumnNotFound {
                column: id_col.to_owned(),
            })?;
        return cast_non_null_utf8(&source_columns[idx], id_col);
    }

    if let Ok(idx) = schema.index_of(COL_NODE_ID) {
        return cast_non_null_utf8(&source_columns[idx], COL_NODE_ID);
    }

    let prefix = options.id_prefix.as_deref().unwrap_or("row");
    let values = (0..rows).map(|idx| format!("{prefix}_{idx}"));
    Ok(Arc::new(StringArray::from_iter_values(values)))
}

fn build_label_array(
    schema: &ArrowSchema,
    source_columns: &[ArrayRef],
    rows: usize,
    options: &CsvNodeReadOptions,
) -> Result<ArrayRef> {
    if let Some(label) = options.label.as_deref() {
        return Ok(Arc::new(repeated_label_array(label, rows)));
    }

    let idx = schema
        .index_of(COL_NODE_LABEL)
        .map_err(|_| GFError::MissingReservedColumn {
            column: COL_NODE_LABEL.to_owned(),
        })?;
    let labels = cast_non_null_utf8(&source_columns[idx], COL_NODE_LABEL)?;
    let labels = labels
        .as_any()
        .downcast_ref::<StringArray>()
        .expect("cast_non_null_utf8 returns StringArray");
    Ok(Arc::new(singleton_label_array(labels)))
}

fn cast_non_null_utf8(array: &ArrayRef, name: &str) -> Result<ArrayRef> {
    if array.null_count() > 0 {
        return Err(GFError::ReservedColumnType {
            column: name.to_owned(),
            expected: "non-null Utf8".to_owned(),
            actual: format!("{:?} with nulls", array.data_type()),
        });
    }

    arrow::compute::cast(array.as_ref(), &DataType::Utf8).map_err(|_| GFError::InvalidCast {
        from: format!("{:?}", array.data_type()),
        to: format!("Utf8 ({name})"),
    })
}

fn repeated_label_array(label: &str, rows: usize) -> arrow::array::ListArray {
    let mut builder = ListBuilder::new(StringBuilder::new());
    for _ in 0..rows {
        builder.values().append_value(label);
        builder.append(true);
    }
    builder.finish()
}

fn singleton_label_array(labels: &StringArray) -> arrow::array::ListArray {
    let mut builder = ListBuilder::new(StringBuilder::new());
    for label in labels.iter() {
        builder
            .values()
            .append_value(label.expect("validated non-null _label"));
        builder.append(true);
    }
    builder.finish()
}

#[cfg(test)]
mod tests {
    use std::io::Write;

    use super::*;

    fn temp_csv(name: &str, contents: &str) -> std::path::PathBuf {
        let mut path = std::env::temp_dir();
        path.push(format!(
            "lynxes_csv_test_{}_{}.csv",
            name,
            std::process::id()
        ));
        let mut file = File::create(&path).unwrap();
        file.write_all(contents.as_bytes()).unwrap();
        path
    }

    #[test]
    fn synthetic_id_and_label() {
        let path = temp_csv("synthetic", "title,year\nMoon,2009\nArrival,2016\n");
        let frame = read_csv_nodes(
            &path,
            &CsvNodeReadOptions {
                label: Some("RawMovie".to_owned()),
                id_prefix: Some("raw_movie".to_owned()),
                ..Default::default()
            },
        )
        .unwrap();

        let ids: Vec<_> = frame.id_column().iter().map(|id| id.unwrap()).collect();
        assert_eq!(ids, ["raw_movie_0", "raw_movie_1"]);
        std::fs::remove_file(path).unwrap();
    }

    #[test]
    fn id_col_is_cast_to_reserved_id() {
        let path = temp_csv("id_col", "id,title\n10,Alien\n11,Aliens\n");
        let frame = read_csv_nodes(
            &path,
            &CsvNodeReadOptions {
                label: Some("RawMovie".to_owned()),
                id_col: Some("id".to_owned()),
                ..Default::default()
            },
        )
        .unwrap();

        let ids: Vec<_> = frame.id_column().iter().map(|id| id.unwrap()).collect();
        assert_eq!(ids, ["10", "11"]);
        assert!(frame.column("id").is_some());
        std::fs::remove_file(path).unwrap();
    }

    #[test]
    fn projection_pushdown_keeps_id_dependency() {
        let path = temp_csv(
            "projection",
            "id,title,cast,votes\n10,Alien,\"[{\"\"name\"\":\"\"Sigourney Weaver\"\"}]\",100\n11,Aliens,\"[]\",200\n",
        );
        let frame = read_csv_nodes(
            &path,
            &CsvNodeReadOptions {
                label: Some("RawMovie".to_owned()),
                id_col: Some("id".to_owned()),
                columns: Some(vec!["title".to_owned()]),
                ..Default::default()
            },
        )
        .unwrap();

        assert_eq!(frame.column_names(), ["_id", "_label", "title"]);
        let ids: Vec<_> = frame.id_column().iter().map(|id| id.unwrap()).collect();
        assert_eq!(ids, ["10", "11"]);
        assert!(frame.column("cast").is_none());
        assert!(frame.column("votes").is_none());
        assert!(frame.column("id").is_none());
        std::fs::remove_file(path).unwrap();
    }

    #[test]
    fn schema_override_can_read_string_view_payload() {
        let path = temp_csv(
            "string_view",
            "id,title,cast\n10,Alien,\"[{\"\"id\"\":1,\"\"name\"\":\"\"Ripley\"\"}]\"\n11,Aliens,\"[]\"\n",
        );
        let frame = read_csv_nodes(
            &path,
            &CsvNodeReadOptions {
                label: Some("RawMovie".to_owned()),
                id_col: Some("id".to_owned()),
                columns: Some(vec!["cast".to_owned()]),
                schema_overrides: vec![("cast".to_owned(), DataType::Utf8View)],
                ..Default::default()
            },
        )
        .unwrap();

        let cast = frame.column("cast").unwrap();
        assert_eq!(cast.data_type(), &DataType::Utf8View);
        assert_eq!(frame.column_names(), ["_id", "_label", "cast"]);
        std::fs::remove_file(path).unwrap();
    }
}
